port/osx/omrosdump.c

```diff
@@ -190,23 +193,34 @@ coredump_to_file(mach_port_t task_port, pid_t pid)
 
 	seg_file_off = file_off + segment_count * sizeof(struct segment_command_64);
 	for (i = 0; i < segment_count; i++) {
-		vm_offset_t data_read = 0;
-		mach_msg_type_number_t data_size = 0;
-
-		kr = mach_vm_read(task_port, segments[i].vmaddr, segments[i].vmsize, &data_read, &data_size);
-		/* any memory segment that is unreadable will be added to the core as an empty segment */
-		if (KERN_SUCCESS == kr) {
-			if ((0 == data_read) || (0 == data_size)) {
-				written = 0;
+		mach_vm_address_t data_read = 0;
+		mach_vm_size_t data_size = 0;
+		mach_vm_size_t bytes_read = 0;
+		data_read = (mach_vm_address_t)seg_data_buffer;
+
+		while(bytes_read < segments[i].vmsize) {
+			mach_vm_size_t to_read = segments[i].vmsize - bytes_read;
+			if (to_read > SEG_DATA_BUFFER_SIZE) {
+				to_read = SEG_DATA_BUFFER_SIZE;
+			}
+			data_size = to_read;
+			kr = mach_vm_read_overwrite(task_port, segments[i].vmaddr + bytes_read, to_read, data_read, &data_size);
+			/* any memory segment that is unreadable will be added to the core as an empty segment */
+			if (KERN_SUCCESS != kr) {
+				break;
+			}
+			if ((0 == data_read) || (to_read != data_size)) {
```

Comment:
How could `0 == data_read` ever be true?

Reply:
Left over from when the code was using `mach_vm_read` instead of the `mach_vm_read_overwrite` which takes an input array now. Will remove.